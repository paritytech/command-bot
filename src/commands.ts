import assert from "assert"
import { Mutex } from "async-mutex"
import fs from "fs"
import glob from "glob"
import path from "path"

import { config } from "src/config"
import { CmdJson } from "src/schema/schema.cmd"
import { CommandRunner } from "src/shell"
import { CommandConfigs } from "src/types"

const CMD_ROOT_FOLDER = "commands"

export async function fetchCommandsConfiguration(devBranch?: string): Promise<CommandConfigs> {
  const cmdRunner = new CommandRunner()
  const scriptsFolder = "scripts"
  const scriptsPath = path.join(config.dataPath, scriptsFolder)

  const commandConfigMutex = new Mutex()

  return await commandConfigMutex.runExclusive<CommandConfigs>(async () => {
    await cmdRunner.run("mkdir", ["-p", scriptsPath])

    if (devBranch && /([^\w\d\-_/]+)/g.test(devBranch)) {
      throw new Error(`Scripts branch should match pattern /([^\\w\\d\\-_/]+)/, given: "${devBranch}", does not match`)
    }

    const scriptsBranch = devBranch || "HEAD"
    let scriptsRevision = await cmdRunner.run("git", [
      "ls-remote",
      `${config.pipelineScripts.repository}`,
      scriptsBranch,
    ])
    if (scriptsRevision instanceof Error) {
      throw scriptsRevision
    }

    if (!scriptsRevision) {
      throw new Error(`Can't find a revision of ${config.pipelineScripts.repository}#${scriptsBranch}`)
    }

    // grab only revision
    scriptsRevision = scriptsRevision
      .trim()
      .split("\t")
      .filter((rev) => !!rev)[0]

    const scriptsRevPath = path.join(scriptsPath, scriptsRevision)
    const commandsRootPath = path.join(scriptsRevPath, CMD_ROOT_FOLDER)
    const commandsOutputPath = path.join(scriptsRevPath, "commands.json")
    const scriptsBranchForClone = devBranch ? ["--branch", devBranch] : []

    if (!fs.existsSync(scriptsRevPath) || !fs.existsSync(commandsOutputPath)) {
      await cmdRunner.run(
        "git",
        [
          "clone",
          "--quiet",
          "--depth",
          "1",
          ...scriptsBranchForClone,
          `${config.pipelineScripts.repository}`,
          scriptsRevPath,
        ],
        { testAllowedErrorMessage: (err) => err.endsWith("already exists and is not an empty directory.") },
      )

      const files: string[] = glob.sync("**/*.cmd.json", { cwd: commandsRootPath })

      const commandConfig: CommandConfigs = files.reduce((configs, file) => {
        // infer command name from the filename
        const cmdName = file.replace(/.*\/([\w-_]+)\.cmd\.json/, "$1")
        const commandPath = path.join(commandsRootPath, file)
        // parse command file contents and save to stack
        configs[cmdName] = JSON.parse(fs.readFileSync(commandPath, "utf8")) as CmdJson

        const cfg = configs[cmdName]?.command.configuration
        if (!cfg.commandStart) {
          cfg.commandStart = getCommandStartConfig(cmdName)
        }
        assert(cfg.commandStart, "CommandStart should be set")

        return configs
      }, {} as CommandConfigs)

      fs.writeFileSync(commandsOutputPath, JSON.stringify(commandConfig))
    }

    return JSON.parse(fs.readFileSync(commandsOutputPath).toString()) as CommandConfigs
  })
}

function getCommandStartConfig(commandName: string): string[] {
  return [`"$PIPELINE_SCRIPTS_DIR/${CMD_ROOT_FOLDER}/${commandName}/${commandName}.sh"`]
}
