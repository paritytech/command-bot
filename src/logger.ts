import { Logger, LoggerOptions } from "opstooling-js"
import { Logger as ProbotLogger } from "pino"
import { getLog } from "probot/lib/helpers/get-log"

export const logFormat = ((): "json" | null => {
  const value = process.env.LOG_FORMAT
  switch (value) {
    case "json": {
      return value
    }
    case undefined: {
      return null
    }
    default: {
      throw new Error(`Invalid $LOG_FORMAT: ${value}`)
    }
  }
})()

export const minLogLevel = ((): "debug" | "info" | "warn" | "error" => {
  const value: string | undefined = process.env.MIN_LOG_LEVEL
  switch (value) {
    case undefined: {
      return "info"
    }
    case "debug":
    case "info":
    case "warn":
    case "error": {
      return value
    }
    default: {
      throw new Error(`Invalid $MIN_LOG_LEVEL: ${value}`)
    }
  }
})()

const loggerOptions: LoggerOptions = { name: "command-bot", minLogLevel, logFormat, impl: console }

export const logger = new Logger(loggerOptions)

export let probotLogger: ProbotLogger | undefined = undefined
switch (logFormat) {
  case "json": {
    probotLogger = getLog({ level: "error", logFormat: "json", logLevelInString: true, logMessageKey: "msg" })
    break
  }
  case null: {
    break
  }
  default: {
    const exhaustivenessCheck: never = logFormat
    // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
    throw new Error(`Not exhaustive: ${exhaustivenessCheck}`)
  }
}
