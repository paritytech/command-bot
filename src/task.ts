import assert from "assert"
import { Mutex } from "async-mutex"
import cp from "child_process"
import { randomUUID } from "crypto"
import { parseISO } from "date-fns"
import EventEmitter from "events"
import LevelErrors from "level-errors"
import { extractRequestError, MatrixClient } from "matrix-bot-sdk"
import { Probot } from "probot"

import { getSortedTasks } from "src/db"

import { getApiTaskEndpoint } from "./api"
import {
  botPullRequestCommentMention,
  botPullRequestCommentSubcommands,
} from "./bot"
import { prepareBranch } from "./core"
import { getPostPullRequestResult, updateComment } from "./github"
import {
  cancelGitlabPipeline,
  restoreTaskGitlabContext,
  runCommandInGitlabPipeline,
} from "./gitlab"
import { Logger } from "./logger"
import { CommandOutput, Context, GitRef } from "./types"
import { displayError, getNextUniqueIncrementalId, intoError } from "./utils"

export const queuedTasks: Map<string, EventEmitter> = new Map()
export const taskExecutionTerminationEvent = Symbol()

export type TaskGitlabPipeline = {
  id: number
  projectId: number
  jobWebUrl: string
}
type TaskBase<T> = {
  tag: T
  id: string
  queuedDate: string
  timesRequeued: number
  timesRequeuedSnapshotBeforeExecution: number
  timesExecuted: number
  gitRef: GitRef
  repoPath: string
  requester: string
  gitlab: {
    job: {
      tags: string[]
      image: string
      variables: Record<string, string | boolean | number>
    }
    pipeline: TaskGitlabPipeline | null
  }
  command: string
}

export type PullRequestTask = TaskBase<"PullRequestTask"> & {
  comment: {
    id: number
    htmlUrl: string
  }
  installationId: number
  gitRef: GitRef & { prNumber: number }
}

export type ApiTask = TaskBase<"ApiTask"> & {
  matrixRoom: string
}

export type Task = PullRequestTask | ApiTask

export const getNextTaskId = () => {
  return `${getNextUniqueIncrementalId()}-${randomUUID()}`
}

export const serializeTaskQueuedDate = (date: Date) => {
  return date.toISOString()
}

export const parseTaskQueuedDate = (str: string) => {
  return parseISO(str)
}

/*
  A Mutex is necessary because otherwise operations from two different commands
  could be interleaved in the same repository, thus leading to
  undefined/unwanted behavior.
  TODO: The Mutex could be per-repository instead of a single one for all
  repositories for better throughput.
*/
const tasksRepositoryLockMutex = new Mutex()

export const queueTask = async (
  parentCtx: Context,
  task: Task,
  {
    onResult,
    updateProgress,
  }: {
    onResult: (result: CommandOutput) => Promise<unknown>
    updateProgress: ((message: string) => Promise<unknown>) | null
  },
) => {
  assert(
    !queuedTasks.has(task.id),
    `Attempted to queue task ${task.id} when it's already registered in the taskMap`,
  )
  const taskEventChannel = new EventEmitter()
  queuedTasks.set(task.id, taskEventChannel)

  const ctx = {
    ...parentCtx,
    logger: parentCtx.logger.child({ taskId: task.id }),
  }
  const { logger, taskDb, getFetchEndpoint, gitlab } = ctx
  const { db } = taskDb

  await db.put(task.id, JSON.stringify(task))

  let terminateTaskExecution: (() => Promise<unknown>) | undefined = undefined
  let activeProcess: cp.ChildProcess | undefined = undefined
  let taskIsAlive = true
  const terminate = async () => {
    if (terminateTaskExecution) {
      await terminateTaskExecution()
      terminateTaskExecution = undefined
      taskEventChannel.emit(taskExecutionTerminationEvent)
    }

    taskIsAlive = false

    queuedTasks.delete(task.id)

    await db.del(task.id)

    if (activeProcess !== undefined) {
      activeProcess.kill()
      logger.info(`Killed child with PID ${activeProcess.pid ?? "?"}`)
      activeProcess = undefined
    }
  }

  const afterTaskRun = (result: CommandOutput | null) => {
    const wasAlive = taskIsAlive

    void terminate().catch((error) => {
      logger.error(error, "Failed to terminate task on afterTaskRun")
    })

    if (wasAlive && result !== null) {
      void onResult(result)
    }
  }

  const additionalTaskCancelInstructions = (() => {
    switch (task.tag) {
      case "PullRequestTask": {
        return `\n\nComment \`${botPullRequestCommentMention} ${botPullRequestCommentSubcommands.cancel}\` to cancel all commands for this pull request.`
      }
      case "ApiTask": {
        return `Send a DELETE request to ${getApiTaskEndpoint(
          task,
        )} for cancelling this task.`
      }
      default: {
        const exhaustivenessCheck: never = task
        // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
        throw new Error(`Not exhaustive: ${exhaustivenessCheck}`)
      }
    }
  })()

  const cancelledMessage = "Task was cancelled"

  void tasksRepositoryLockMutex
    .runExclusive(async () => {
      try {
        await db.put(
          task.id,
          JSON.stringify({
            ...task,
            timesRequeuedSnapshotBeforeExecution: task.timesRequeued,
            timesExecuted: task.timesExecuted + 1,
          }),
        )

        const restoredTaskGitlabCtx = await restoreTaskGitlabContext(ctx, task)
        if (restoredTaskGitlabCtx !== undefined) {
          return restoredTaskGitlabCtx
        }

        if (taskIsAlive) {
          logger.info(task, "Starting task")
        } else {
          logger.info(task, "Task was cancelled before it could start")
          return cancelledMessage
        }

        const prepareBranchSteps = prepareBranch(ctx, task, {
          getFetchEndpoint: () => {
            return getFetchEndpoint(
              "installationId" in task ? task.installationId : null,
            )
          },
        })
        while (taskIsAlive) {
          const next = await prepareBranchSteps.next()
          if (next.done) {
            break
          }

          activeProcess = undefined

          if (typeof next.value !== "string") {
            return next.value
          }
        }
        if (!taskIsAlive) {
          return cancelledMessage
        }

        const pipelineCtx = await runCommandInGitlabPipeline(ctx, task)

        task.gitlab.pipeline = {
          id: pipelineCtx.id,
          jobWebUrl: pipelineCtx.jobWebUrl,
          projectId: pipelineCtx.projectId,
        }
        await db.put(task.id, JSON.stringify(task))

        if (updateProgress) {
          await updateProgress(
            `@${task.requester} ${pipelineCtx.jobWebUrl} was started for your command \`${task.command}\`. Check out https://${gitlab.domain}/${gitlab.pushNamespace}/${task.gitRef.repo}/-/pipelines?page=1&scope=all&username=${gitlab.accessTokenUsername} to know what else is being executed currently. ${additionalTaskCancelInstructions}`,
          )
        }

        return pipelineCtx
      } catch (error) {
        return intoError(error)
      }
    })
    .then((taskPipeline) => {
      if (
        taskPipeline instanceof Error ||
        typeof taskPipeline === "string" ||
        taskPipeline === null
      ) {
        return afterTaskRun(taskPipeline)
      }

      terminateTaskExecution = taskPipeline.terminate

      taskPipeline
        .waitUntilFinished(taskEventChannel)
        .then(() => {
          afterTaskRun(
            `${taskPipeline.jobWebUrl} has ${
              taskIsAlive ? "finished" : "was cancelled"
            }. If any artifacts were generated, you can download them from ${
              taskPipeline.jobWebUrl
            }/artifacts/download.`,
          )
        })
        .catch(afterTaskRun)
    })
    .catch(afterTaskRun)

  return `${task.command} was queued. ${additionalTaskCancelInstructions}`
}

export const requeueUnterminatedTasks = async (ctx: Context, bot: Probot) => {
  const { taskDb, logger, matrix } = ctx
  const { db } = taskDb

  /*
    unterminatedItems are leftover tasks from previous server instances which
    were not finished properly for some reason (e.g. the server was restarted).
  */
  const unterminatedItems = await getSortedTasks(ctx, { onlyNotAlive: true })

  for (const {
    task: { timesRequeued, ...task },
    id,
  } of unterminatedItems) {
    await db.del(id)

    const prepareRequeuedTask = <T>(prevTask: T) => {
      logger.info(prevTask, "Prepare requeue")
      return { ...prevTask, timesRequeued: timesRequeued + 1 }
    }

    type RequeueComponent = {
      requeue: () => Promise<unknown> | unknown
      announceCancel: (msg: string) => Promise<unknown> | unknown
    }
    const getRequeueResult = async (): Promise<RequeueComponent | Error> => {
      try {
        switch (task.tag) {
          case "PullRequestTask": {
            const {
              gitRef: { owner, repo, prNumber: prNumber },
              comment,
              requester,
            } = task

            const octokit = await bot.auth(task.installationId)

            const announceCancel = (message: string) => {
              return updateComment(ctx, octokit, {
                owner,
                repo,
                pull_number: prNumber,
                comment_id: comment.id,
                body: `@${requester} ${message}`,
              })
            }

            const requeuedTask = prepareRequeuedTask(task)
            const requeue = () => {
              return queueTask(ctx, requeuedTask, {
                onResult: getPostPullRequestResult(ctx, octokit, requeuedTask),
                /*
                  Assumes the relevant progress update was already sent when
                  the task was queued for the first time, thus there's no need
                  to keep updating it
                  TODO: Update the item in the database to tell when
                  updateProgress no longer needs to be called.
                */
                updateProgress: null,
              })
            }

            return { requeue, announceCancel }
          }
          case "ApiTask": {
            if (matrix === null) {
              return {
                announceCancel: () => {
                  logger.warn(
                    task,
                    "ApiTask cannot be requeued because Matrix client is missing",
                  )
                },
                requeue: () => {},
              }
            }

            const { matrixRoom } = task
            const sendMatrixMessage = (msg: string) => {
              return matrix.sendText(matrixRoom, msg)
            }

            const requeuedTask = prepareRequeuedTask(task)
            return {
              announceCancel: sendMatrixMessage,
              requeue: () => {
                return queueTask(ctx, requeuedTask, {
                  onResult: getSendTaskMatrixResult(
                    matrix,
                    logger,
                    requeuedTask,
                  ),
                  /*
                    Assumes the relevant progress update was already sent when
                    the task was queued for the first time, thus there's no need
                    to keep updating it
                    TODO: Update the item in the database to tell when
                    updateProgress no longer needs to be called.
                  */
                  updateProgress: null,
                })
              },
            }
          }
          default: {
            const exhaustivenessCheck: never = task
            // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
            throw new Error(`Not exhaustive: ${exhaustivenessCheck}`)
          }
        }
      } catch (error) {
        return intoError(error)
      }
    }

    const requeueResult = await getRequeueResult()
    if (requeueResult instanceof Error) {
      logger.fatal(requeueResult, "Exception while trying to requeue a task")
      continue
    }

    const { announceCancel, requeue } = requeueResult
    if (
      timesRequeued &&
      /*
        Check if the task was requeued and got to execute, but it failed for
        some reason, in which case it will not be retried further; in
        comparison, it might have been requeued and not had a chance to execute
        due to other crash-inducing command being in front of it, thus it's not
        reasonable to avoid rescheduling this command if it's not his fault
      */
      timesRequeued === task.timesRequeuedSnapshotBeforeExecution
    ) {
      await announceCancel(
        `Command was rescheduled and failed to finish (check for task id ${id} in the logs); execution will not automatically be restarted further.`,
      )
    } else {
      try {
        await requeue()
      } catch (error) {
        const errorMessage = displayError(error)
        await announceCancel(
          `Caught exception while trying to reschedule the command; it will not be rescheduled further. Error message: ${errorMessage}.`,
        )
      }
    }
  }
}

export const getSendTaskMatrixResult = (
  matrix: MatrixClient,
  logger: Logger,
  task: ApiTask,
) => {
  return async (message: CommandOutput) => {
    try {
      await matrix.sendText(
        task.matrixRoom,
        `Task ID ${task.id} has finished with message "${
          message instanceof Error ? displayError(message) : message
        }"`,
      )
    } catch (rawError) {
      const error = intoError(rawError)
      logger.error(
        extractRequestError(error),
        "Caught error when sending Matrix message",
      )
    }
  }
}

export const cancelTask = async (ctx: Context, taskId: Task | string) => {
  const {
    taskDb: { db },
    logger,
  } = ctx

  const task =
    typeof taskId === "string"
      ? await (async () => {
          try {
            return JSON.parse(await db.get(taskId)) as Task
          } catch (error) {
            if (error instanceof LevelErrors.NotFoundError) {
              return error
            } else {
              throw error
            }
          }
        })()
      : taskId
  if (task instanceof Error) {
    return task
  }

  logger.info(task, "Cancelling task")

  if (task.gitlab.pipeline !== null) {
    await cancelGitlabPipeline(ctx, task.gitlab.pipeline)
  }

  await db.del(task.id)
}
