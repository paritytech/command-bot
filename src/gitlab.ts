import EventEmitter from "events"
import Joi from "joi"
import fetch from "node-fetch"
import path from "path"
import yaml from "yaml"

import { CommandRunner, fsWriteFile } from "./shell"
import { Task, taskExecutionTerminationEvent, TaskGitlabPipeline } from "./task"
import { Context } from "./types"
import { validatedFetch } from "./utils"

export const runCommandInGitlabPipeline = async (ctx: Context, task: Task) => {
  const { logger, gitlab } = ctx

  const cmdRunner = new CommandRunner(ctx, {
    itemsToRedact: [gitlab.accessToken],
    shouldTrackProgress: false,
    cwd: task.repoPath,
  })

  /*
    Save the head SHA before doing any modifications to the branch so that
    scripts will be able to restore the branch as it was on GitHub
  */
  const headSha = await cmdRunner.run("git", ["rev-parse", "HEAD"])

  const artifactsFolderPath = ".git/.artifacts"

  const getPipelineScriptsCloneCommand = ({
    withRef,
  }: {
    withRef: boolean
  }) => {
    return `git clone --depth 1 ${
      withRef ? `--branch "$PIPELINE_SCRIPTS_REF"` : ""
    } "$PIPELINE_SCRIPTS_REPOSITORY" "$PIPELINE_SCRIPTS_DIR"`
  }

  await fsWriteFile(
    path.join(task.repoPath, ".gitlab-ci.yml"),
    yaml.stringify({
      workflow: {
        rules: [
          { if: `$CI_PIPELINE_SOURCE == "api"` },
          { if: `$CI_PIPELINE_SOURCE == "web"` },
        ],
      },
      command: {
        ...task.gitlab.job,
        script: [
          // prettier-ignore
          'if [ "${PIPELINE_SCRIPTS_REPOSITORY:-}" ]; then ' +
            'if [ "${PIPELINE_SCRIPTS_REF:-}" ]; then ' +
              getPipelineScriptsCloneCommand({ withRef: true }) + "; " +
            "else " +
              getPipelineScriptsCloneCommand({ withRef: false }) + "; " +
            "fi" + "; " +
          "fi",
          `export ARTIFACTS_DIR="$PWD/${artifactsFolderPath}"`,
          `mkdir -p "$ARTIFACTS_DIR"`,
          task.command,
        ],
        artifacts: {
          name: "${CI_JOB_NAME}_${CI_COMMIT_REF_NAME}",
          expire_in: "7 days",
          when: "always",
          paths: [artifactsFolderPath],
        },
        variables: {
          ...task.gitlab.job.variables,
          GH_CONTRIBUTOR: task.gitRef.contributor,
          GH_CONTRIBUTOR_REPO: task.gitRef.repo,
          GH_CONTRIBUTOR_BRANCH: task.gitRef.branch,
          GH_HEAD_SHA: headSha,
          COMMIT_MESSAGE: task.command,
          PIPELINE_SCRIPTS_REPOSITORY: ctx.pipelineScripts?.repository,
          PIPELINE_SCRIPTS_REF: ctx.pipelineScripts?.ref,
          PIPELINE_SCRIPTS_DIR: ".git/.scripts",
        },
      },
    }),
  )

  const branchName = `cmd-bot/${
    "prNumber" in task.gitRef ? task.gitRef.prNumber : task.gitRef.branch
  }`
  await cmdRunner.run("git", ["branch", "-D", branchName], {
    testAllowedErrorMessage: (err) => {
      return err.endsWith("not found.")
    },
  })
  await cmdRunner.run("git", ["checkout", "-b", branchName])

  await cmdRunner.run("git", ["add", ".gitlab-ci.yml"])

  await cmdRunner.run("git", ["commit", "-m", task.command])

  const gitlabRemote = "gitlab"
  const gitlabProjectPath = `${gitlab.pushNamespace}/${task.gitRef.repo}`

  await cmdRunner.run("git", ["remote", "remove", gitlabRemote], {
    testAllowedErrorMessage: (err) => {
      return err.includes("No such remote:")
    },
  })

  await cmdRunner.run("git", [
    "remote",
    "add",
    gitlabRemote,
    `https://token:${gitlab.accessToken}@${gitlab.domain}/${gitlabProjectPath}.git`,
  ])

  /*
    It's not necessary to say "--option ci.skip" because the pipeline execution
    is conditional per workflow:rules
  */
  await cmdRunner.run("git", ["push", "--force", gitlabRemote, "HEAD"])

  const pipeline = await validatedFetch<{
    id: number
    project_id: number
  }>(
    fetch(
      `https://${gitlab.domain}/api/v4/projects/${encodeURIComponent(
        gitlabProjectPath,
      )}/pipeline?ref=${encodeURIComponent(branchName)}`,
      { method: "POST", headers: { "PRIVATE-TOKEN": gitlab.accessToken } },
    ),
    Joi.object()
      .keys({
        id: Joi.number().required(),
        project_id: Joi.number().required(),
      })
      .options({ allowUnknown: true }),
  )
  logger.info(pipeline, `Created pipeline for task ${task.id}`)

  const [job] = await validatedFetch<
    [
      {
        web_url: string
      },
    ]
  >(
    fetch(
      `https://${gitlab.domain}/api/v4/projects/${pipeline.project_id}/pipelines/${pipeline.id}/jobs`,
      { headers: { "PRIVATE-TOKEN": gitlab.accessToken } },
    ),
    Joi.array()
      .items(
        Joi.object()
          .keys({ web_url: Joi.string().required() })
          .options({ allowUnknown: true }),
      )
      .length(1)
      .required(),
  )
  logger.info(job, `Fetched job for task ${task.id}`)

  return getAliveTaskGitlabContext(ctx, {
    id: pipeline.id,
    projectId: pipeline.project_id,
    jobWebUrl: job.web_url,
  })
}

export const cancelGitlabPipeline = async (
  { gitlab, logger }: Context,
  pipeline: TaskGitlabPipeline,
) => {
  logger.info(pipeline, "Cancelling GitLab pipeline")
  await validatedFetch(
    fetch(
      /*
        Note: this endpoint can be called any time, even if the pipeline has
        already finished
      */
      `https://${gitlab.domain}/api/v4/projects/${pipeline.projectId}/pipelines/${pipeline.id}/cancel`,
      { method: "POST", headers: { "PRIVATE-TOKEN": gitlab.accessToken } },
    ),
    Joi.object()
      .keys({ id: Joi.number().required() })
      .options({ allowUnknown: true }),
  )
}

const isPipelineFinished = async (
  ctx: Context,
  pipeline: TaskGitlabPipeline,
) => {
  const { gitlab } = ctx

  const { status } = await validatedFetch<{
    status: string
  }>(
    fetch(
      `https://${gitlab.domain}/api/v4/projects/${pipeline.projectId}/pipelines/${pipeline.id}`,
      { headers: { "PRIVATE-TOKEN": gitlab.accessToken } },
    ),
    Joi.object()
      .keys({ status: Joi.string().required() })
      .options({ allowUnknown: true }),
  )
  switch (status) {
    case "success":
    case "skipped":
    case "canceled":
    case "failed": {
      return true
    }
  }
}

export const restoreTaskGitlabContext = async (ctx: Context, task: Task) => {
  const { pipeline } = task.gitlab
  if (!pipeline) {
    return
  }

  if (await isPipelineFinished(ctx, pipeline)) {
    return null
  }

  return getAliveTaskGitlabContext(ctx, pipeline)
}

const getAliveTaskGitlabContext = (
  ctx: Context,
  pipeline: TaskGitlabPipeline,
): TaskGitlabPipeline & {
  terminate: () => Promise<Error | undefined>
  waitUntilFinished: (taskEventChannel: EventEmitter) => Promise<unknown>
} => {
  const { logger } = ctx

  let wasTerminated = false
  return {
    ...pipeline,
    terminate: async () => {
      wasTerminated = true
      await cancelGitlabPipeline(ctx, pipeline)
      return undefined
    },
    waitUntilFinished: (taskEventChannel) => {
      return Promise.race([
        new Promise<void>((resolve) => {
          taskEventChannel.on(taskExecutionTerminationEvent, () => {
            wasTerminated = true
            resolve()
          })
        }),
        new Promise<void>((resolve, reject) => {
          /*
            Avoid potentially cancelling a costly long-running job if the GitLab
            request fails randomly
          */
          let subsequentErrors: unknown[] = []
          const pollPipelineCompletion = async () => {
            if (wasTerminated) {
              return resolve()
            }
            try {
              if (await isPipelineFinished(ctx, pipeline)) {
                return resolve()
              }
              subsequentErrors = []
            } catch (error) {
              subsequentErrors.push(error)
              if (subsequentErrors.length > 2) {
                logger.error(
                  { errors: subsequentErrors, pipeline },
                  `GitLab pipeline status polling failed ${subsequentErrors.length} in a row. Aborting...`,
                )
                return reject(error)
              }
            }
            setTimeout(() => {
              void pollPipelineCompletion()
            }, 16384)
          }
          void pollPipelineCompletion()
        }),
      ])
    },
  }
}
