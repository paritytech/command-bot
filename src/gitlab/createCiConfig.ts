import { Task } from "src/task"
import { PipelineScripts } from "src/types"

export function createCiConfig(
  headSha: string,
  task: Task,
  pipelineScripts: PipelineScripts,
  jobTaskInfoMessage: string,
): object {
  const artifactsFolderPath = ".git/.artifacts"
  return {
    workflow: { rules: [{ if: `$CI_PIPELINE_SOURCE == "api"` }, { if: `$CI_PIPELINE_SOURCE == "web"` }] },
    command: {
      timeout: "24 hours",
      ...task.gitlab.job,
      script: [
        ...`
        echo "This job is related to task ${task.id}. ${jobTaskInfoMessage}."

        # The scripts repository might be left over from a previous run in the
        # same Gitlab shell executor
        rm -rf "$PIPELINE_SCRIPTS_DIR"

        if [ "\${PIPELINE_SCRIPTS_REPOSITORY:-}" ]; then
          if [ "\${PIPELINE_SCRIPTS_REF:-}" ]; then
            git clone --progress --verbose --depth 1 --branch "$PIPELINE_SCRIPTS_REF" "$PIPELINE_SCRIPTS_REPOSITORY" "$PIPELINE_SCRIPTS_DIR"
          else
            git clone --progress --verbose --depth 1 "$PIPELINE_SCRIPTS_REPOSITORY" "$PIPELINE_SCRIPTS_DIR"
          fi
        fi
        export ARTIFACTS_DIR="$PWD/${artifactsFolderPath}"
        # The artifacts directory might be left over from a previous run in
        # the same Gitlab shell executor
        rm -rf "$ARTIFACTS_DIR"
        mkdir -p "$ARTIFACTS_DIR"
        `.split("\n"),
        task.command,
      ],
      artifacts: {
        name: "${CI_JOB_NAME}_${CI_COMMIT_REF_NAME}",
        expire_in: "7 days",
        when: "always",
        paths: [artifactsFolderPath],
      },
      variables: {
        // overrideable variables
        PIPELINE_SCRIPTS_REF: pipelineScripts?.ref,
        ...task.gitlab.job.variables,
        // non-overrideable variables
        GH_OWNER: task.gitRef.upstream.owner,
        GH_OWNER_REPO: task.gitRef.upstream.repo,
        ...(task.gitRef.upstream.branch ? { GH_OWNER_BRANCH: task.gitRef.upstream.branch } : {}),
        GH_CONTRIBUTOR: task.gitRef.contributor.owner,
        GH_CONTRIBUTOR_REPO: task.gitRef.contributor.repo,
        GH_CONTRIBUTOR_BRANCH: task.gitRef.contributor.branch,
        GH_HEAD_SHA: headSha,
        COMMIT_MESSAGE: task.command,
        PIPELINE_SCRIPTS_REPOSITORY: pipelineScripts?.repository,
        PIPELINE_SCRIPTS_DIR: ".git/.scripts",
      },
    },
  }
}
