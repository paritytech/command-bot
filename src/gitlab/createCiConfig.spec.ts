import { jest } from "@jest/globals"

import { createCiConfig } from "src/gitlab/createCiConfig"
import { logger } from "src/logger"
import { Task } from "src/task"

jest.mock("src/commands")

logger.options.minLogLevel = "fatal"

function getTaskStub(opts: { vars: Record<string, string> } = { vars: {} }): Task {
  return {
    gitlab: {
      job: { tags: ["any"], variables: opts.vars, image: "image" },
      pipeline: { id: 12, jobWebUrl: "jobWebUrl", projectId: 99 },
    },
    gitRef: {
      upstream: { owner: "upstream_owner", repo: "upstream_repo", branch: "upstream_branch" },
      contributor: { owner: "contributor_owner", repo: "contributor_repo", branch: "contributor_branch" },
      prNumber: 123,
    },
    command: "echo 123",
    tag: "PullRequestTask",
    id: "123",
    queuedDate: "queuedDate",
    timesRequeued: 0,
    timesRequeuedSnapshotBeforeExecution: 0,
    timesExecuted: 0,
    repoPath: "repoPath",
    requester: "requester",
    comment: { id: 55, htmlUrl: "htmlUrl" },
    installationId: 888,
  }
}

describe("createCiConfig", () => {
  test(`PIPELINE_SCRIPTS_REF and other VARs are applied, while PIPELINE_SCRIPTS_REPOSITORY is not overrideable`, () => {
    const res = createCiConfig(
      "headSha",
      getTaskStub({
        vars: {
          CUSTOM_VAR: "IS_APPLIED",
          PIPELINE_SCRIPTS_REF: "OVERRIDDEN_REF_IS_APPLIED",
          PIPELINE_SCRIPTS_REPOSITORY: "OVERRIDDEN_REPO_NOT_APPLIED",
        },
      }),
      { ref: "OVERRIDABLE", repository: "NON_OVERRIDEABLE" },
      "job Task info message",
    )
    expect(res).toMatchSnapshot()
  })
})
