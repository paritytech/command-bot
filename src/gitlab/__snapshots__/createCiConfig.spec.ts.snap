// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createCiConfig PIPELINE_SCRIPTS_REF and other VARs are applied, while PIPELINE_SCRIPTS_REPOSITORY is not overrideable 1`] = `
{
  "command": {
    "artifacts": {
      "expire_in": "7 days",
      "name": "\${CI_JOB_NAME}_\${CI_COMMIT_REF_NAME}",
      "paths": [
        ".git/.artifacts",
      ],
      "when": "always",
    },
    "image": "image",
    "script": [
      "",
      "        echo "This job is related to task 123. job Task info message."",
      "",
      "        # The scripts repository might be left over from a previous run in the",
      "        # same Gitlab shell executor",
      "        rm -rf "$PIPELINE_SCRIPTS_DIR"",
      "",
      "        if [ "\${PIPELINE_SCRIPTS_REPOSITORY:-}" ]; then",
      "          if [ "\${PIPELINE_SCRIPTS_REF:-}" ]; then",
      "            git clone --progress --verbose --depth 1 --branch "$PIPELINE_SCRIPTS_REF" "$PIPELINE_SCRIPTS_REPOSITORY" "$PIPELINE_SCRIPTS_DIR"",
      "          else",
      "            git clone --progress --verbose --depth 1 "$PIPELINE_SCRIPTS_REPOSITORY" "$PIPELINE_SCRIPTS_DIR"",
      "          fi",
      "        fi",
      "        export ARTIFACTS_DIR="$PWD/.git/.artifacts"",
      "        # The artifacts directory might be left over from a previous run in",
      "        # the same Gitlab shell executor",
      "        rm -rf "$ARTIFACTS_DIR"",
      "        mkdir -p "$ARTIFACTS_DIR"",
      "        ",
      "echo 123",
    ],
    "tags": [
      "any",
    ],
    "timeout": "24 hours",
    "variables": {
      "COMMIT_MESSAGE": "echo 123",
      "CUSTOM_VAR": "IS_APPLIED",
      "GH_CONTRIBUTOR": "contributor_owner",
      "GH_CONTRIBUTOR_BRANCH": "contributor_branch",
      "GH_CONTRIBUTOR_REPO": "contributor_repo",
      "GH_HEAD_SHA": "headSha",
      "GH_OWNER": "upstream_owner",
      "GH_OWNER_BRANCH": "upstream_branch",
      "GH_OWNER_REPO": "upstream_repo",
      "PIPELINE_SCRIPTS_DIR": ".git/.scripts",
      "PIPELINE_SCRIPTS_REF": "OVERRIDDEN_REF_IS_APPLIED",
      "PIPELINE_SCRIPTS_REPOSITORY": "NON_OVERRIDEABLE",
    },
  },
  "workflow": {
    "rules": [
      {
        "if": "$CI_PIPELINE_SOURCE == "api"",
      },
      {
        "if": "$CI_PIPELINE_SOURCE == "web"",
      },
    ],
  },
}
`;
